### ISSUE TITLE
Fix false-negatives regarding LITERAL_ELSE & LITERAL_CATCH to detect K&R Blocks in google_checks.xml

### ISSUE THREAD

**Issue by Zopsss** [checkstyle/checkstyle#15791]
I have read check documentation: https://checkstyle.org/checks/blocks/leftcurly.html https://checkstyle.org/checks/blocks/rightcurly.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

follow up of https://github.com/checkstyle/checkstyle/issues/15664

In the above mentioned issue, `LITERAL_FINALLY` was moved from `RightCurlySame` to `RightCurlyAlone` as part of this PR: https://github.com/checkstyle/checkstyle/pull/15789

Now the current behavior of google_checks.xml for detecting K & R blocks is as follows:

From google_checks.xml:
```xml
    <module name="RightCurly">
      <property name="id" value="RightCurlySame"/>
      <property name="tokens"
               value="LITERAL_TRY, LITERAL_CATCH, LITERAL_IF, LITERAL_ELSE,
                    LITERAL_DO"/>
    </module>
    <module name="RightCurly">
      <property name="id" value="RightCurlyAlone"/>
      <property name="option" value="alone"/>
      <property name="tokens"
               value="CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE, STATIC_INIT,
                    INSTANCE_INIT, ANNOTATION_DEF, ENUM_DEF, INTERFACE_DEF, RECORD_DEF,
                    COMPACT_CTOR_DEF, LITERAL_SWITCH, LITERAL_CASE, LITERAL_FINALLY"/>
    </module>
    <module name="SuppressionXpathSingleFilter">
      <!-- suppression is required till https://github.com/checkstyle/checkstyle/issues/7541 -->
      <property name="id" value="RightCurlyAlone"/>
      <property name="query" value="//RCURLY[parent::SLIST[count(./*)=1]
                                     or preceding-sibling::*[last()][self::LCURLY]]"/>
    </module>
```

```java
package com.google.checkstyle.test.chapter4formatting.rule412nonemptyblocks;

/** some javadoc. */
public class InputTryCatchIfElse {

  @interface TesterAnnotation {} //ok

  void foo() throws Exception {
    int a = 90;
    boolean test = true;

    if (a == 1) {
    } else {}

    if (a == 1) {
    } else { }

    if (a == 45) {}

    if (a == 9) {} else {} // false-negative

    if (a == 99) {
      System.out.println("test");
    }
    else {
      System.out.println("before");
    }

    try (MyResource r = new MyResource()) { }

    try (MyResource r = new MyResource()) {}

    try (MyResource r = new MyResource()) {} catch (Exception expected) {} // false-negative

    try (MyResource r = new MyResource()) {} catch (Exception expected) { } // false-negative

    try (MyResource r = new MyResource()) {
    } catch (Exception expected) {} // false-negative

    try (MyResource r = new MyResource()) {
    } catch (Exception expected) { } // false-negative

    try (MyResource r = new MyResource()) { ; }

    try {
      /* foo */
    }
    catch (NullPointerException e) {
      /* foo */
    }
    catch (Exception e) {
      /* foo */
    }
    finally {
      test = true;
    }

    try {
      /* foo */
    } catch (NullPointerException e) {
      /* foo */
    }
    catch (Exception e) {
      /* foo */
    }
    finally {
      test = true;
    }

    try {
      /* foo */
    }
    catch (Exception e) {
      /* foo */
    }
    finally {
      test = true;
    }

    try {
      /* foo */
    } catch (Exception e) {
      /* foo */
    }
    finally {
      test = true;
    }
  }

  /** some javadoc. */
  public class MyResource implements AutoCloseable {
    /** some javadoc. */
    @Override
    public void close() throws Exception {
      System.out.println("Closed MyResource");
    }
  }
}
```

```
$ java -jar .\checkstyle-10.19.0-SNAPSHOT-all.jar -c .\google_checks.xml .\InputTryCatchIfElse.java
Starting audit...
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:16: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:24:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:29: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:35: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:41: Empty blocks should have no spaces. Empty blocks                                    may only be represented as {} when not part of a                                    multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:43:43: '{' at column 43 should have line break after. [LeftCurlyEol]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:47:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:50:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:53:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:62:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:65:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:72:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:75:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse.java:84:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
Audit done.
```

```java
package com.google.checkstyle.test.chapter4formatting.rule412nonemptyblocks;

/** some javadoc. */
public class InputTryCatchIfElse2 {
  /** some javadoc. */
  public static void main(String[] args) {
    boolean after = false;

    try {
      /* foo */
    } catch (NullPointerException e) {
      /* foo */
    } catch (Exception e) {
      /* foo */
    } finally {
      after = true;
    }

    if (after) {
      System.out.println("after");
    } else if (after) {
      System.out.println("before");
    } else if (after) {
      System.out.println("before");
    } else {
      System.out.println("before");
    }

    if (after) {
      System.out.println("after");
    } else if (after) {} else if (after) { // false-negative
      System.out.println("before");
    } else if (!after) { /* foo */ } else { // false-negative
      System.out.println("before");
    }

    if (after) {
      System.out.println("foo");
    } else if (false) {
      System.out.println("foo");
    } else {
      System.out.println("foo");
    }

    if (after) {
      System.out.println("after");
    }
    else if (after) {
      System.out.println("before");
    }
    else {
      System.out.println("before");
    }

    if (after) {
      System.out.println("after");
    }
    else if (after) {
      System.out.println("before");
    }
    else if (after) {
      System.out.println("before");
    }
    else {
      System.out.println("before");
    }

    if (after) {
      System.out.println("after");
    }
    else if (after) {
      System.out.println("before");
    } else {
      System.out.println("before");
    }

    if (after) {
      System.out.println("after");
    } else if (after) {
      System.out.println("before");
    }
    else {
      System.out.println("before");
    }
  }
}
```

```
$ java -jar .\checkstyle-10.19.0-SNAPSHOT-all.jar -c .\google_checks.xml .\InputTryCatchIfElse2.java
Starting audit...
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse2.java:47:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse2.java:50:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse2.java:57:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse2.java:60:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse2.java:63:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse2.java:70:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
[WARN] C:\checkstyle testing\.\InputTryCatchIfElse2.java:81:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurlySame]
Audit done.
```
Both of these input files are taken from K&R block input files.

As part of this issue, we need to fix all false-negatives. Generate diff reports and find edge cases, fix them too.

The issue will most probably require modifying suppression for `LITERAL_ELSE` & `LITERAL_CATCH`.

**Comment by Karthikk-18** [comment]
@romani, @Zopsss I'm on it

**Comment by DakshRJain737** [comment]
@Karthikk-18 Are you still working on this issue?? 
Can i take over if you are no longer continuing with this PR.

**Comment by DakshRJain737** [comment]
@Karthikk-18 ?????

**Comment by Zopsss** [comment]
@DakshRJain737 you can start working on the issue if you want. The other contributor's PR seems to be closed.

**Comment by DakshRJain737** [comment]
> [@DakshRJain737](https://github.com/DakshRJain737) you can start working on the issue if you want. The other contributor's PR seems to be closed.

Sure, I have already started working. Will issue a PR when all my pending PRs gets merged

**Comment by Youssefalaa7** [comment]
@DakshRJain737 are you still working on this issue or i can take over?

**Comment by DakshRJain737** [comment]
> [@DakshRJain737](https://github.com/DakshRJain737) are you still working on this issue or i can take over?

@Youssefalaa7 I was working on this but I am waiting for my other PRs to merge before I initiate a new PR

You can work on this issue. No problem at all.

**Comment by Youssefalaa7** [comment]
@DakshRJain737 thanks, I am on it

**Comment by romani** [comment]
```
$ cat Test.java
package com.google.checkstyle.test.chapter4formatting.rule412nonemptyblocks;

/** some javadoc. */
public class Test {

  @interface TesterAnnotation {} //ok

  void foo() throws Exception {
    int a = 90;
    boolean test = true;

    if (a == 1) {
    } else {}

    if (a == 1) {
    } else { }

    if (a == 45) {}

    if (a == 9) {} else {} // false-negative

    if (a == 99) {
      System.out.println("test");
    }
    else {
      System.out.println("before");
    }
  }
}

$ java -jar google-java-format-1.35.0-all-deps.jar Test.java > FormattedTest.java
$ cat FormattedTest.java 
package com.google.checkstyle.test.chapter4formatting.rule412nonemptyblocks;

/** some javadoc. */
public class Test {

  @interface TesterAnnotation {} // ok

  void foo() throws Exception {
    int a = 90;
    boolean test = true;

    if (a == 1) {
    } else {
    }

    if (a == 1) {
    } else {
    }

    if (a == 45) {}

    if (a == 9) {
    } else {
    } // false-negative

    if (a == 99) {
      System.out.println("test");
    } else {
      System.out.println("before");
    }
  }
}

$ diff -u Test.java FormattedTest.java
--- Test.java	2026-07-23 21:21:19.387880499 -0700
+++ FormattedTest.java	2026-07-23 21:24:43.761900014 -0700
@@ -3,26 +3,29 @@
 /** some javadoc. */
 public class Test {
 
-  @interface TesterAnnotation {} //ok
+  @interface TesterAnnotation {} // ok
 
   void foo() throws Exception {
     int a = 90;
     boolean test = true;
 
     if (a == 1) {
-    } else {}
+    } else {
+    }
 
     if (a == 1) {
-    } else { }
+    } else {
+    }
 
     if (a == 45) {}
 
-    if (a == 9) {} else {} // false-negative
+    if (a == 9) {
+    } else {
+    } // false-negative
 
     if (a == 99) {
       System.out.println("test");
-    }
-    else {
+    } else {
       System.out.println("before");
     }
   }

```

**Comment by romani** [comment]
```
$ cat Test.java
package com.google.checkstyle.test.chapter4formatting.rule412nonemptyblocks;

/** some javadoc. */
public class Test {
  /** some javadoc. */
  public static void main(String[] args) {
    boolean after = false;

    try {
      /* foo */
    } catch (NullPointerException e) {
      /* foo */
    } catch (Exception e) {
      /* foo */
    } finally {
      after = true;
    }

    if (after) {
      System.out.println("after");
    } else if (after) {
      System.out.println("before");
    } else if (after) {
      System.out.println("before");
    } else {
      System.out.println("before");
    }

    if (after) {
      System.out.println("after");
    } else if (after) {} else if (after) { // false-negative
      System.out.println("before");
    } else if (!after) { /* foo */ } else { // false-negative
      System.out.println("before");
    }
  }

$ java -jar google-java-format-1.35.0-all-deps.jar Test.java > FormattedTest.java
$ cat FormattedTest.java 
package com.google.checkstyle.test.chapter4formatting.rule412nonemptyblocks;

/** some javadoc. */
public class Test {
  /** some javadoc. */
  public static void main(String[] args) {
    boolean after = false;

    try {
      /* foo */
    } catch (NullPointerException e) {
      /* foo */
    } catch (Exception e) {
      /* foo */
    } finally {
      after = true;
    }

    if (after) {
      System.out.println("after");
    } else if (after) {
      System.out.println("before");
    } else if (after) {
      System.out.println("before");
    } else {
      System.out.println("before");
    }

    if (after) {
      System.out.println("after");
    } else if (after) {
    } else if (after) { // false-negative
      System.out.println("before");
    } else if (!after) {
      /* foo */
    } else { // false-negative
      System.out.println("before");
    }
  }
}

$ diff -u Test.java FormattedTest.java
--- Test.java	2026-07-23 21:28:00.873668031 -0700
+++ FormattedTest.java	2026-07-23 21:28:17.243852927 -0700
@@ -28,10 +28,13 @@
 
     if (after) {
       System.out.println("after");
-    } else if (after) {} else if (after) { // false-negative
+    } else if (after) {
+    } else if (after) { // false-negative
       System.out.println("before");
-    } else if (!after) { /* foo */ } else { // false-negative
+    } else if (!after) {
+      /* foo */
+    } else { // false-negative
       System.out.println("before");
     }
   }
```

**Comment by romani** [comment]
General purpose Check will not work
We need to make new Check as described at https://github.com/checkstyle/checkstyle/issues/7541

@Zopsss , are you ok to close this issue ?
